crates/sui/src/ptb/ptb_parser/errors.rs

```diff
@@ -31,6 +33,289 @@ macro_rules! err {
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct FileIndexedErrors(pub BTreeMap<(String, usize), Vec<PTBCommand>>);
+
+impl FileIndexedErrors {
+    /// Take a set of commands and index them by file scope.
+    pub fn new(commands: &BTreeMap<usize, PTBCommand>) -> Self {
+        let mut file_indexed_commands = BTreeMap::new();
+        let mut name_collision_count: BTreeMap<_, _> =
+            [("console".to_owned(), 0)].into_iter().collect();
+        let mut scope_stack = vec![];
+        let mut current_scope = ("console".to_string(), 0);
+        for (_, command) in commands {
+            if command.name == FILE_START {
+                // Push a dummy command to keep indices in line with the usage of the `--file`
+                // command.
+                file_indexed_commands
+                    .entry(current_scope.clone())
+                    .or_insert_with(Vec::new)
+                    .push(command.clone());
+                scope_stack.push(current_scope.clone());
+                let name_index = name_collision_count
+                    .entry(command.values[0].clone())
+                    .and_modify(|i| *i += 1)
+                    .or_insert(0);
+                current_scope = (command.values[0].clone(), *name_index);
+                continue;
+            }
+
+            if command.name == FILE_END {
+                current_scope = scope_stack.pop().unwrap();
+                continue;
+            }
+
+            file_indexed_commands
+                .entry(current_scope.clone())
+                .or_insert_with(Vec::new)
+                .push(command.clone());
+        }
+
+        Self(file_indexed_commands)
+    }
+
+    pub fn get(&self, file_scope: &FileScope) -> Option<&PTBCommand> {
+        self.0
+            .get(&(file_scope.name.clone(), file_scope.name_index))
+            .and_then(|commands| commands.get(file_scope.file_command_index))
+    }
+}
+
+impl Span {
+    pub fn new(start: usize, end: usize, arg_idx: usize) -> Self {
+        Self {
+            start,
+            end,
+            arg_idx,
+        }
+    }
+
+    /// Return the union of this span with a set of other spans.
+    pub fn union_with(&self, others: impl IntoIterator<Item = Span>) -> Span {
+        let mut start = self.start;
+        let mut end = self.end;
+
+        for s in others {
+            start = start.min(s.start);
+            end = end.max(s.end);
+        }
+        Span {
+            start,
+            end,
+            arg_idx: self.arg_idx,
+        }
+    }
+}
+
+pub fn span<T: Debug + Clone + PartialEq + Eq>(loc: Span, value: T) -> Spanned<T> {
+    Spanned { span: loc, value }
+}
+
+// If no span we point to the command name
+// If there is a span, we convert the span range to the appropriate offset in the whole string for
+// the command
+fn convert_span(
+    original_command: PTBCommand,
+    error: PTBError,
+) -> (String, LabeledSpan, String, Option<String>) {
```

Comment:
Might be better to return a type than a 4-tuple, but there might be reasons as to why that's not desirable, so will let you decide the best way.

Reply:
Nah, just laziness/since it's a one-off lol. happy to make it into a type :)